File: Functions/extractResistanceValues.m
function [checkupResistance_OhmTimeStamp, checkupResistance_Ohm,checkupResistenceFEC] = extractResistanceValues(timeWithGaps, voltage, current, timeS, startTime, endTime, cellNum, cellLabel)
% extractResistanceValues - Calculate internal resistance from current pulses
%
% This function identifies high-current pulse segments and calculates DC
% resistance from the voltage drop during pulses using Ohm's law.
%
% Inputs:
%   timeWithGaps - Datetime array with timestamps (may contain NaT)
%   voltage      - Voltage array in Volts
%   current      - Current array in Amperes
%   timeS        - Time array in seconds (may contain NaN values)
%   startTime    - Start datetime for analysis range
%   endTime      - End datetime for analysis range
%   cellNum      - Cell identifier string for plot title
%   cellLabel    - (Optional) Descriptive label from test plan
%
% Outputs:
%   checkupResistanceTimeStamp - Datetime array of resistance measurement timestamps
%   checkupResistance_Ohm          - Array of measured DC resistance values (Ω)

if nargin < 8
    cellLabel = '';
end

fprintf('\nExtracting resistance values from current pulses...\n');
tic;  % Restart timer

% Select data within the specified datetime range
selectedIndices = (timeWithGaps >= startTime) & (timeWithGaps <= endTime);
selectedTime = timeWithGaps(selectedIndices);
selectedVoltage = voltage(selectedIndices);
selectedCurrent = current(selectedIndices);
selectedTimeS = timeS(selectedIndices);

% Identify high-current pulse segments
constantCurrentValue = -58;  % Target current in Amperes
tolerance = 0.4;             % Tolerance in Amperes
constantCurrentIndices = abs(selectedCurrent - constantCurrentValue) <= tolerance;
AverageLength = 5; %number of samples to average V and I

% Filter for pulse segments (30-300 data points, typically ~30s duration)
minSegmentLength = 30; % to be deleted
maxSegmentLength = 300; % to be deleted
minSegmentTime_s = 29;
maxSegementTime_s = 31;

% Find segments meeting length criteria
segments = findSegmentsMinMax(constantCurrentIndices, AverageLength, selectedTime, minSegmentTime_s,maxSegementTime_s);
fprintf('Found %d current pulse segments. (Elapsed: %.2f s)\n', length(segments), toc);
if isempty(segments)
    warning('no segments found')
    checkupResistance_OhmTimeStamp=[];
    checkupResistance_Ohm=[];
    checkupResistenceFEC=[];
    return
end

BatteryCapacity_Ah = 58;
FullEquivalentCycles = cumtrapz(selectedTimeS,abs(selectedCurrent))/BatteryCapacity_Ah/3600/2;

% Initialize figure for resistance analysis
figure('Units','normalized', 'OuterPosition',[0 0 0.5 1]); % Create a figure that takes up half the screen width


% Define line styles for plot variation
lineStyles = {'-', '--', ':', '-.'};

% Loop through each pulse segment and calculate resistance
for i = 1:length(segments)
    if contains(cellNum,'A3') || contains(cellNum,'A4')
        if i==1 % skip the first test of cell A3 and A4. 
            continue
        end
    end

    % Select line style for this iteration
    lineIdx = mod(i-1, length(lineStyles)) + 1;
    currentLineStyle = lineStyles{lineIdx};
    
    segmentIndices = segments{i};
    % Include one point before pulse to capture voltage drop
    segmentIndices = [segmentIndices(1)-1 segmentIndices];
    segmentVoltage = selectedVoltage(segmentIndices);
    segmentCurrent = selectedCurrent(segmentIndices);
    segmentTime = selectedTime(segmentIndices);
    segmentTimeS = selectedTimeS(segmentIndices);
    segmentTimeS = segmentTimeS - segmentTimeS(1);  % Normalize to start at 0

    % this fixes an error on the original data. While concatenating
    % datasets, an artificial delayed was introduced in some of these
    % test
    if segmentTimeS(AverageLength*2+2)-segmentTimeS(AverageLength*2+1)>2 
        samplingTime_S = segmentTimeS(3)-segmentTimeS(2);
        DeltaTime_s = segmentTimeS(AverageLength*2+2)-segmentTimeS(AverageLength*2+1);
        segmentTimeS(AverageLength*2+2:end) = segmentTimeS(AverageLength*2+2:end) - DeltaTime_s + samplingTime_S;
        warning('At time %s, the algorithm corrected a delay created by concatenating multiple files of the original test data',string(segmentTime(1)));
    end
    
    % Plot voltage response to pulse
    subplot(2,1,1)
    hold on;
    plot(segmentTimeS, segmentVoltage, ...
         'LineStyle', currentLineStyle, 'LineWidth', 1.5, ...
         'Color', [i/length(segments), 0, 1-i/length(segments)]);
    xlabel('Time [s]');
    ylabel('Voltage [V]');
    
    % Plot current pulse
    subplot(2,1,2)
    hold on;
    plot(segmentTimeS, segmentCurrent, ...
         'LineStyle', currentLineStyle, 'LineWidth', 1.5, ...
         'Color', [i/length(segments), 0, 1-i/length(segments)]);
    xlabel('Time [s]');
    ylabel('Current [A]');
    
    % Calculate DC resistance from ohm's law: R = ΔV / ΔI
    checkupResistance_OhmTimeStamp(i) = segmentTime(1);
    DeltaVoltage_V = mean(segmentVoltage(end-AverageLength:end)) - mean(segmentVoltage(1:AverageLength));
    DeltaCurrent_A = mean(segmentCurrent(end-AverageLength:end)) - mean(segmentCurrent(1:AverageLength));
    checkupResistance_Ohm(i) = (DeltaVoltage_V) / (DeltaCurrent_A);
    checkupResistenceFEC(i) = FullEquivalentCycles(segmentIndices(1));
end
if isempty(cellLabel)
    sgtitle(['30s Resistance Analysis - V Response to I Pulses ' cellNum],'interpreter','none');
else
    sgtitle(['30s Resistance Analysis - V Response to I Pulses ' cellNum ' - ' cellLabel],'interpreter','none');
end
lgd = legend(string(checkupResistance_OhmTimeStamp));
lgd.Location = 'best';
fprintf('Resistance extraction complete. (Elapsed: %.2f s)\n', toc);

% Remove NaT entries from output arrays
validIdx = ~isnat(checkupResistance_OhmTimeStamp);
checkupResistance_OhmTimeStamp = checkupResistance_OhmTimeStamp(validIdx);
checkupResistance_Ohm = checkupResistance_Ohm(validIdx);
checkupResistenceFEC = checkupResistenceFEC(validIdx);

end

function segments = findSegmentsMinMax(indices, AverageLength, selectedTime, minDuration_s,maxDuration_S)
    % Find continuous segments where indices are true, with min and max length
    %
    % Inputs:
    %   indices   - Logical array indicating which points meet criteria
    %   minLength - Minimum number of consecutive true values
    %   maxLength - Maximum number of consecutive true values
    %
    % Outputs:
    %   segments  - Cell array where each cell contains index ranges
    
    segments = {};
    startIdx = find(indices, 1, 'first');
    
    while ~isempty(startIdx)
        % Find where the segment ends
        endIdx = find(~indices(startIdx:end), 1, 'first') + startIdx - 2;
        if isempty(endIdx)
            endIdx = length(indices);
        end
        
        % Only keep segments within length bounds
        segmentLength = endIdx - startIdx + 1;

        Duration_s = max(selectedTime(startIdx:endIdx))-min(selectedTime(startIdx:endIdx));

          if Duration_s>=seconds(minDuration_s)
              if Duration_s<=seconds(maxDuration_S)
                   segments{end+1} = startIdx-AverageLength*2:endIdx;   
              end
          end
%         if segmentLength >= minLength && Duration_s>seconds(minDuration_s)
%             if segmentLength <= maxLength
%                 segments{end+1} = startIdx:endIdx;         
%             else
%                 endIdxTmp = startIdx + maxLength;
%                 segments{end+1} = startIdx:endIdxTmp; 
%             end
%         end
        
        % Find next segment start
        startIdx = find(indices(endIdx+1:end), 1, 'first') + endIdx;
    end
end

File: Functions/extractResistanceValues.py
"""
extractResistanceValues - Calculate internal resistance from current pulses

This module identifies high-current pulse segments and calculates DC
resistance from the voltage drop during pulses using Ohm's law.

Author: Converted from MATLAB to Python
Date: November 25, 2025
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
try:  # scipy >= 1.6 renamed cumtrapz; it was removed in newer releases.
    from scipy.integrate import cumulative_trapezoid as cumtrapz
except ImportError:
    from scipy.integrate import cumtrapz
import time
import warnings
import re

from cellNumberToGroupChannel import cell_number_to_group_channel


def extract_resistance_values(time_with_gaps, voltage, current, time_s, 
                              start_time, end_time, cell_num, cell_label=''):
    """
    Calculate internal resistance from current pulses.
    
    This function identifies high-current pulse segments and calculates DC
    resistance from the voltage drop during pulses using Ohm's law.
    
    Parameters
    ----------
    time_with_gaps : numpy.ndarray
        Datetime array with timestamps (may contain NaT)
    voltage : numpy.ndarray
        Voltage array in Volts
    current : numpy.ndarray
        Current array in Amperes
    time_s : numpy.ndarray
        Time array in seconds (may contain NaN values)
    start_time : datetime or np.datetime64
        Start datetime for analysis range
    end_time : datetime or np.datetime64
        End datetime for analysis range
    cell_num : str
        Cell identifier string for plot title
    cell_label : str, optional
        Descriptive label from test plan (default: '')
    
    Returns
    -------
    checkup_resistance_timestamp : numpy.ndarray
        Datetime array of resistance measurement timestamps
    checkup_resistance : numpy.ndarray
        Array of measured DC resistance values (Ω)
    checkup_resistance_fec : numpy.ndarray
        Array of FEC values at resistance measurements
    """
    
    print('\nExtracting resistance values from current pulses...')
    tic = time.time()
    
    # Select data within the specified datetime range
    selected_indices = (time_with_gaps >= start_time) & (time_with_gaps <= end_time)
    selected_time = time_with_gaps[selected_indices]
    selected_voltage = voltage[selected_indices]
    selected_current = current[selected_indices]
    selected_time_s = time_s[selected_indices]
    
    # Identify high-current pulse segments
    constant_current_value = -58  # Target current in Amperes
    tolerance = 0.4               # Tolerance in Amperes
    constant_current_indices = np.abs(selected_current - constant_current_value) <= tolerance
    average_length = 5  # Number of samples to average V and I
    
    # Filter for pulse segments by time duration (typically ~30s duration)
    min_segment_time_s = 29
    max_segment_time_s = 31
    
    # Find segments meeting duration criteria
    segments = _find_segments_min_max(constant_current_indices, average_length, 
                                       selected_time, min_segment_time_s, max_segment_time_s)
    
    elapsed = time.time() - tic
    print(f'Found {len(segments)} current pulse segments. (Elapsed: {elapsed:.2f} s)')
    
    # Return empty arrays if no segments found
    if len(segments) == 0:
        warnings.warn('no segments found')
        return np.array([]), np.array([]), np.array([])
    
    # Calculate Full Equivalent Cycles using cumtrapz over time (to match MATLAB)
    # Use the already time-range-filtered selected_current/selected_time_s (matches
    # MATLAB's FullEquivalentCycles = cumtrapz(selectedTimeS, abs(selectedCurrent))/...);
    # using the raw, unfiltered time_s here would propagate NaN gap markers through
    # the whole cumulative integral once cumtrapz crosses the first gap.
    battery_capacity_ah = 58
    selected_time_s = np.asarray(selected_time_s)
    abs_current = np.abs(np.nan_to_num(selected_current))
    full_equivalent_cycles = np.concatenate(([0], cumtrapz(abs_current, selected_time_s))) / battery_capacity_ah / 3600 / 2
    
    # Initialize output arrays
    checkup_resistance_timestamp = []
    checkup_resistance = []
    checkup_resistance_fec = []
    
    # Initialize figure for resistance analysis (half screen width, full height)
    fig, (ax1, ax2) = plt.subplots(2, 1)
    # Get screen size and set figure to half width, full height
    fig_manager = plt.get_current_fig_manager()
    try:
        # Try to set window geometry (works with TkAgg, Qt backends)
        screen_width = fig.canvas.manager.window.winfo_screenwidth()
        screen_height = fig.canvas.manager.window.winfo_screenheight()
        fig_manager.window.geometry(f'{screen_width//2}x{screen_height}+0+0')
    except AttributeError:
        try:
            # For Qt backend
            fig_manager.window.setGeometry(0, 0, 
                                           fig_manager.window.screen().size().width() // 2,
                                           fig_manager.window.screen().size().height())
        except AttributeError:
            # Fallback: use figure size in inches
            fig.set_size_inches(10, 12)
    
    # Define line styles for plot variation
    line_styles = ['-', '--', ':', '-.']

    # Determine whether this cell belongs to the A3 (45 degC) or A4 (0-45 degC)
    # campaign, via the cell-number crosswalk (R-025: cell_num no longer carries
    # the campaign letter as a substring, so it can't be string-matched anymore).
    cell_digits = re.search(r'Cell_(\d+)', cell_num)
    is_campaign_a3_or_a4 = False
    if cell_digits:
        temp_group, _ = cell_number_to_group_channel(int(cell_digits.group(1)))
        is_campaign_a3_or_a4 = temp_group in (3, 4)

    # Loop through each pulse segment and calculate resistance
    for i, segment_indices in enumerate(segments):
        # Skip the first test for cell A3 and A4
        if is_campaign_a3_or_a4:
            if i == 0:
                continue
        
        # Select line style for this iteration
        line_idx = i % len(line_styles)
        current_line_style = line_styles[line_idx]
        
        # Include one point before pulse to capture voltage drop
        segment_indices = np.concatenate([[segment_indices[0] - 1], segment_indices])
        
        segment_voltage = selected_voltage[segment_indices]
        segment_current = selected_current[segment_indices]
        segment_time = selected_time[segment_indices]
        segment_time_s_local = selected_time_s[segment_indices].copy()
        segment_time_s_local = segment_time_s_local - segment_time_s_local[0]  # Normalize to start at 0
        
        # Fix for artificial delay introduced during data concatenation
        check_idx = average_length * 2 + 1
        if len(segment_time_s_local) > check_idx + 1:
            if segment_time_s_local[check_idx + 1] - segment_time_s_local[check_idx] > 2:
                sampling_time_s = segment_time_s_local[2] - segment_time_s_local[1]
                delta_time_s = segment_time_s_local[check_idx + 1] - segment_time_s_local[check_idx]
                segment_time_s_local[check_idx + 1:] = segment_time_s_local[check_idx + 1:] - delta_time_s + sampling_time_s
                warnings.warn(f'At time {segment_time[0]}, the algorithm corrected a delay created by concatenating multiple files of the original test data')
        
        # Color based on segment position (gradient from blue to red)
        num_segments = len(segments)
        color = [i / num_segments, 0, 1 - i / num_segments]
        
        # Plot voltage response to pulse
        ax1.plot(segment_time_s_local, segment_voltage,
                linestyle=current_line_style, linewidth=1.5, color=color)
        
        # Plot current pulse
        ax2.plot(segment_time_s_local, segment_current,
                linestyle=current_line_style, linewidth=1.5, color=color)
        
        # Calculate DC resistance from Ohm's law: R = ΔV / ΔI
        # Use averaged values at start and end of segment
        checkup_resistance_timestamp.append(segment_time[0])
        delta_v = np.mean(segment_voltage[-average_length:]) - np.mean(segment_voltage[:average_length])
        delta_i = np.mean(segment_current[-average_length:]) - np.mean(segment_current[:average_length])
        resistance = delta_v / delta_i if delta_i != 0 else np.nan
        checkup_resistance.append(resistance)
        checkup_resistance_fec.append(full_equivalent_cycles[segment_indices[0]])
    
    # Format subplot 1
    ax1.set_xlabel('Time [s]')
    ax1.set_ylabel('Voltage [V]')
    ax1.grid(True, alpha=0.3)
    
    # Format subplot 2
    ax2.set_xlabel('Time [s]')
    ax2.set_ylabel('Current [A]')
    ax2.grid(True, alpha=0.3)
    
    # Add title to figure
    if cell_label:
        fig.suptitle(f'30s Resistance Analysis - V Response to I Pulses {cell_num} - {cell_label}')
    else:
        fig.suptitle(f'30s Resistance Analysis - V Response to I Pulses {cell_num}')
    
    # Add legend with timestamps
    if len(checkup_resistance_timestamp) > 0:
        legend_labels = [str(ts) for ts in checkup_resistance_timestamp]
        ax1.legend(legend_labels, fontsize=8, loc='best')
    
    plt.tight_layout()
    plt.show(block=False)
    plt.pause(0.001)  # Brief pause to ensure plot renders
    
    elapsed = time.time() - tic
    print(f'Resistance extraction complete. (Elapsed: {elapsed:.2f} s)')
    
    # Convert lists to numpy arrays
    checkup_resistance_timestamp = np.array(checkup_resistance_timestamp)
    checkup_resistance = np.array(checkup_resistance)
    checkup_resistance_fec = np.array(checkup_resistance_fec)
    
    # Remove NaT entries from output arrays
    valid_idx = pd.notna(checkup_resistance_timestamp)
    checkup_resistance_timestamp = checkup_resistance_timestamp[valid_idx]
    checkup_resistance = checkup_resistance[valid_idx]
    checkup_resistance_fec = checkup_resistance_fec[valid_idx]
    
    return checkup_resistance_timestamp, checkup_resistance, checkup_resistance_fec


def _find_segments_min_max(indices, average_length, selected_time, min_duration_s, max_duration_s):
    """
    Find continuous segments where indices are True, with duration constraints.
    
    Parameters
    ----------
    indices : numpy.ndarray
        Boolean array indicating which points meet criteria
    average_length : int
        Number of samples to include before segment for averaging
    selected_time : numpy.ndarray
        Datetime array for duration calculation
    min_duration_s : float
        Minimum segment duration in seconds
    max_duration_s : float
        Maximum segment duration in seconds
    
    Returns
    -------
    segments : list of numpy.ndarray
        List where each element contains index ranges (as arrays)
    """
    
    segments = []
    
    # Find first True index
    true_indices = np.where(indices)[0]
    if len(true_indices) == 0:
        return segments
    
    start_idx = true_indices[0]
    
    while start_idx is not None:
        # Find where the segment ends
        remaining = indices[start_idx:]
        false_positions = np.where(~remaining)[0]
        
        if len(false_positions) == 0:
            end_idx = len(indices) - 1
        else:
            end_idx = false_positions[0] + start_idx - 1
        
        # Calculate segment duration
        segment_time = selected_time[start_idx:end_idx + 1]
        if len(segment_time) > 0:
            # Handle datetime64 or datetime objects
            try:
                duration = segment_time.max() - segment_time.min()
                # Convert to seconds if timedelta
                if hasattr(duration, 'total_seconds'):
                    duration_s = duration.total_seconds()
                else:
                    # numpy timedelta64
                    duration_s = duration / np.timedelta64(1, 's')
            except:
                duration_s = 0
            
            # Only keep segments within duration bounds
            if min_duration_s <= duration_s <= max_duration_s:
                # Include average_length*2 points before segment
                segment_start = max(0, start_idx - average_length * 2)
                segments.append(np.arange(segment_start, end_idx + 1))
        
        # Find next segment start
        remaining_after = indices[end_idx + 1:]
        next_true = np.where(remaining_after)[0]
        
        if len(next_true) == 0:
            start_idx = None
        else:
            start_idx = next_true[0] + end_idx + 1
    
    return segments


# For compatibility with import statement in main script
__all__ = ['extract_resistance_values']